File: 973.k-closest-points-to-origin.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=973 lang=cpp
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> kClosest(vector<vector<int>> &points, int k)
    {
        int x, y;
        vector<vector<int>> ans;
        priority_queue<vector<int>> pq;

        for (auto point : points)
        {
            x = point[0];
            y = point[1];
            pq.push({-((x * x) + (y * y)), x, y});
        }

        while (k--)
        {
            ans.push_back({pq.top()[1], pq.top()[2]});
            pq.pop();
        }

        return ans;
    }
};
// @lc code=end

File: Java/Grind75/Week 3/973.k-closest-points-to-origin.java
import java.util.*;

/*
 * @lc app=leetcode id=973 lang=java
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        if (k == points.length) {
            return points;
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>(
                (p1, p2) -> Double.compare(p2[0] * p2[0] + p2[1] * p2[1],
                        p1[0] * p1[0] + p1[1] * p1[1]));
        for (int[] point : points) {
            pq.add(point);
            if (pq.size() > k) {
                pq.poll();
            }
        }
        return pq.toArray(new int[0][0]);
    }
}
// @lc code=end
